docs/auth/backend-auth.md

```diff
@@ -0,0 +1,82 @@
+---
+id: backend-auth
+title: Secure your backend
+description: Documentation on securing your backend by requiring authentication.
+---
+
+##
+
+By default, the backstage backend does not require authentication. Securing your backend requires additional configuration or a separate service like Google's Identity-Aware Proxy. Since the backend is served as an express app we can leverage express middleware to require JWT bearer tokens for all requests. The `@backstage/plugin-auth-node` package provides several factory methods for creating express middleware that can be used to secure your backend:
+
+- `createRequireJwtTokenMiddleware`: Creates middleware that requires a JWT bearer token for all requests.
+- `createRequireUserTokenMiddleware`: Creates middleware that requires a JWT bearer token for all requests and verifies that the token is a valid user token.
+- `createRequireServerTokenMiddleware`: Creates middleware a JWT bearer token for all requests and verifies that the token is a valid server token (see [service to service Auth](./service-to-service-auth.md)).
+
+> NOTE: `createRequireJwtTokenMiddleware` does not validate the JWT tokens, instead you must specify an express middleware that will be invoked when the token has the specified audience. `createRequireUserTokenMiddleware` and `createRequireServerTokenMiddleware` both validate the tokens.
+
+The middleware created by these factories only verifies that a valid JWT token is provided and thus adds **authentication** to the api endpoint. The middleware cannot check whether the identity enclosed in the token has access to specific resources (**authorisation**). Each plugin backend is responsible for checking the permissions of the identity in the token if required.
+
+To configure all backend routes, except `/auth`, to require a valid user or server token, you can use the following code:
+
+```ts title="packages/backend/src/index.ts"
+// ...
+/* highlight-add-start */
+import {
+  createRequireJwtTokenMiddleware,
+  createRequireServerTokenMiddleware,
+  createRequireUserTokenMiddleware,
+} from '@internal/plugin-auth-node';
+/* highlight-add-end */
+
+// ...
+async function main() {
+  // ...
+  const createEnv = makeCreateEnv(config);
+
+  /* highlight-add-start */
+  const tokenManager = ServerTokenManager.fromConfig(config, {
+    logger: getRootLogger(),
+  });
+  const discovery = SingleHostDiscovery.fromConfig(config);
+  const identity = DefaultIdentityClient.create({
+    discovery,
+  });
+
+  const requireBackstageToken = createRequireJwtTokenMiddleware({
+    backstage: createRequireUserTokenMiddleware(identity),
+    'backstage-server': createRequireServerTokenMiddleware(tokenManager),
+  });
+  /* highlight-add-end */
+
+  const apiRouter = Router();
+
+  /* highlight-add-next-line */
+  // unauthenticated routes
+  apiRouter.use('/auth', await auth(authEnv));
```

Comment:
You don't have the app backend in this example, did you think about that case a bit? I don't think the entire frontend app is something that we can leak, and need to protect that as well. It does mean we need a separate log-in page that is not protected though.

Reply:
No I only thought about the backend/api side of things. Are you talking about front-end resources like javascript bundles? Isn't it common to have those types of resources public?

Reply:
Yep, especially JavaScript. In the case of Backstage it's very common to be building bespoke plugins for internal tech platforms. I think it is common enough to not want to make that publicly available that we need to avoid it, since it leaks information about an organization. Happy to be challenged on that assumption though, and to be honest it could make sense to simply leave that as a gap that is closed with a more sophisticated auth setup.